File: system_equation_solver.py
#!/System/Volumes/Data/anaconda3/bin/python

# -*- coding: utf-8 -*-
# @Author: Andres Nowak
# @Date: Tue Feb 25 13:06:15 CST 2020

# This program uses the Gauss-Jordan method

from fractions import Fraction as frac
from sys import exit


def main():
    print("""the equations you write should look like this:\n
			1x + 2y + 1z = 4\n
			3x + 0y + 1z = 2\n
			1x - 1y + 1z = 1\n
            and the result if its negative like this = -4""")
    size_of_equation = int(input(
        "Of how many unknowns is the system of equations: "))

    matrix_of_equations, list_of_letters_of_unknowns = create_matrix_of_values_of_equations_from_input(
        size_of_equation)

    resolved_matrix = run_through_matrix(matrix_of_equations)

    print_values_of_unknowns(resolved_matrix, list_of_letters_of_unknowns)


def create_matrix_of_values_of_equations_from_input(size_of_equation):
    list_of_letters_of_unknowns = []
    matrix_of_equations = []

    for i in range(size_of_equation):
        equation = input(f"{i + 1} equation: ")
        equation = equation.split()

        equation_values = []

        sign = 1

        for index, value in enumerate(equation):
            if index % 2 == 0:
                if index != len(equation) - 1:
                    equation_values.append(float(value[0:-1]) * sign)

                    list_of_letters_of_unknowns.append(value[-1])
                else:
                    equation_values.append(float(value))
            else:
                if(value != "="):
                    sign = value + "1"
                    sign = int(sign)

        matrix_of_equations.append(equation_values)

    return matrix_of_equations, list_of_letters_of_unknowns


def print_values_of_unknowns(matrix, list_of_letters_of_unknowns):
    for index, row in enumerate(matrix):
        print(f"{list_of_letters_of_unknowns[index]} = {float(row[-1]):.5f}")


def run_through_matrix(matrix):
    matrix_size = len(matrix)
    order_matrix = create_matrix_operation_order(matrix_size)

    for index_c, col in enumerate(order_matrix):
        for row in col:
            if row == index_c:
                convert_to_1(matrix[row], matrix[row][index_c])
            else:
                convert_to_0(matrix[row], matrix[col[0]], matrix[row][index_c])

    return matrix


def create_matrix_operation_order(matrix_size):
    order_matrix = []

    for i in range(matrix_size):
        row = []
        for j in range(matrix_size):
            if j == 0:
                row.append(i)
            if j != i:
                row.append(j)

        order_matrix.append(row)

    return order_matrix


def convert_to_1(row, value):
    try:
        value = frac(value)
        inverse_of_value = value**-1
        for index, i in enumerate(row):
            row[index] = frac(i) * inverse_of_value
    except ZeroDivisionError:
        print("This system of equations cant be solved")
        exit()


def convert_to_0(row, row_to_multiply, value):
    index = 0
    value *= -1

    for i, j in zip(row, row_to_multiply):
        row[index] = (j * value) + i

        index += 1


if __name__ == "__main__":
    """matrix = [[1, -6, 3, -2], [2, -3, 1, -2], [3, 3, -2, 2]]
    resolved_matrix = run_through_matrix(matrix)"""

    main()

File: system_equation_solver.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"sync"
)

func main() {
	fmt.Printf("the equations you write should look like this:\n1x + 2y + 1z = 4\n3x + 0y + 1z = 2\n1x + -1y + 1z = 1\n")

	fmt.Printf("Of how many unknowns is the system of equations: ")
	var size_of_equation_string string
	fmt.Scanln(&size_of_equation_string)

	size_of_equation, _ := strconv.ParseInt(size_of_equation_string, 10, 64)

	matrix_of_equations, list_of_letters_of_unknowns := createMatrixOfValuesOfEquationsFromInput(size_of_equation)

	resolved_matrix := runThroughMatrix(matrix_of_equations)

	printValuesOfUnknowns(resolved_matrix, list_of_letters_of_unknowns)
}

func createMatrixOfValuesOfEquationsFromInput(size_of_equation int64) ([][]float64, []string) {
	list_of_letters_of_unknowns := []string{}
	matrix_of_equations := [][]float64{}

	var i int64
	for i = 0; i < size_of_equation; i++ {
		fmt.Printf("%d equation: ", i+1)

		//Todo: Fix problem where input doesnt work with print from another program
		equation_input := bufio.NewReader(os.Stdin)
		equation_string, _ := equation_input.ReadString('\n')

		equation := strings.Split(equation_string, " ")

		equation_values := []float64{}

		for index, value := range equation {
			if index%2 == 0 {
				if index < len(equation)-1 {
					list_of_letters_of_unknowns = append(list_of_letters_of_unknowns, string(value[len(value)-1]))

					value = value[0 : len(value)-1]
					value, _ := strconv.ParseFloat(value, 64)

					equation_values = append(equation_values, float64(value))
				} else {
					value, _ := strconv.ParseFloat(value[0:len(value)-1], 64)

					equation_values = append(equation_values, value)
				}
			}
		}
		matrix_of_equations = append(matrix_of_equations, equation_values)
	}

	return matrix_of_equations, list_of_letters_of_unknowns
}

func printValuesOfUnknowns(matrix [][]float64, list_of_letters_of_unknowns []string) {
	for index, row := range matrix {
		fmt.Printf("%s = %.5f\n", list_of_letters_of_unknowns[index], row[len(row)-1])
	}
}

func runThroughMatrix(matrix [][]float64) [][]float64 {
	matrix_size := len(matrix)
	order_matrix := createMatrixOperationOrder(matrix_size)

	for index_r, row := range order_matrix {
		for _, col := range row {
			if col == index_r {
				convertTo1(matrix[col], matrix[col][index_r])
			} else {
				convertTo0(matrix[col], matrix[row[0]], matrix[col][index_r])
			}
		}
	}

	return matrix
}

func createMatrixOperationOrder(matrix_size int) [][]int {
	order_matrix := [][]int{}

	for i := 0; i < matrix_size; i++ {
		row := []int{}
		for j := 0; j < matrix_size; j++ {
			if j == 0 {
				row = append(row, i)
			}
			if j != i {
				row = append(row, j)
			}
		}
		order_matrix = append(order_matrix, row)
	}
	return order_matrix
}

func convertTo1(row []float64, value float64) {
	inverse_of_value := math.Pow(value, -1)

	for index, i := range row {
		row[index] = i * inverse_of_value
	}
}

func convertTo0(row []float64, row_to_multiply []float64, value float64) {
	value *= -1

	// This is for multithreading
	var wg sync.WaitGroup

	size := len(row)
	wg.Add(size)

	for index := 0; index < size; index++ {
		go func(index int) {
			defer wg.Done()
			row[index] = (row_to_multiply[index] * value) + row[index]
		}(index)
	}

	wg.Wait()
}
